### ISSUE TITLE
GLM 5.1 pcc drop debug

### ISSUE THREAD

**Issue by ipotkonjak-tt** [tenstorrent/tt-metal#47914]
We're observing drop down to 0.88 pcc on later layers of the model (layer 60 and later). The drop recovers and output token appears to be correct, but the drop needs debugging.

**Comment by nmilicevicTT** [comment]
# GLM-5.1 prefill — PCC investigation (findings, data formats, comparisons, runs)

Bring-up of GLM-5.1 (HF `glm_moe_dsa`: absorbed-MQA MLA + DSA lightning-indexer top-2048 +
256-expert top-8 noaux_tc MoE, 78 layers) on Blackhole Galaxy, tp4 (8x4). This collects everything
we found about PCC: what's validated, the "0.88 chained-PCC" story, the near-degenerate gate, the
data-format/precision map, every lever tried, and how GLM should be validated.

---

## TL;DR — what works and what doesn't

**Works (GLM is functionally correct):**
- **Next-token exact:** device greedy token = `19264 [' calm'] @ 100%`, matches the GPU (vLLM) `completion_token_ids=[19264]`.
- **Teacher-forced (same routing both sides) PCC:** `0.977` (cpu_v32 ref) / `0.987` (both-inject). This is the compute floor and it's high.
- **Every component is faithful:** dense layers L0-2 decoder PCC `0.9995-0.9997`; sparse-SDPA input q `0.9995`; gate logits device↔GPU `0.99997 @L3`; indexer top-2048 device↔vLLM `0.98` (logits `0.9995`); MoE reference == HF `GlmMoeDsaMoE` **PCC 1.0**.
- **Block test (fused decoder layer):** dense `0.99999996` / moe `0.99999964` (real weights); dense `0.9997` / moe `0.9946` (random). Full block vs HF `GlmMoeDsaModel` (real weights) `0.99996`.
- **Op audit:** ttnn MLA/indexer/decoder structure matches HF exactly (only nit was q_a/kv_a norm eps 1e-5→1e-6, numerically inert, fixed).

**Doesn't work (as a *metric*, not as the model):**
- **Free-running chained per-layer PCC vs the vLLM fp8 trace troughs at ~0.88 @ L62** (recovers to ~0.96 @ L77). This is **NOT a device bug** — it's an intrinsic property of GLM's near-degenerate gate amplifying *any* implementation/precision difference. Two correct NVIDIA GPU implementations (HF vs vLLM) hit the *same* 0.88 vs each other; the GPU can't even reproduce its own routing run-to-run.

**Blocked / dead-ends:**
- **bf8 routed experts:** L1 overflow (+4% over 1.5 MB) — a kernel program-config change, not env-gated. Experts stay bf4.
- **Beating ~0.89 chained:** would require fp8-simulating the whole forward, and **fp8-GEMM ≡ bf16** (measured), so fp8-sim is a dead end. bf4 experts cap it anyway.

**Recommendation:** validate GLM by **next-token exact + teacher-forced PCC (0.977/0.987) + component/block PCC (0.9995-1.0)** — NOT by free-running chained-vs-single-trace ≥ 0.95, which is unreachable by any implementation including reference GPUs.

---

## Setup

- Dims: hidden 6144, q_lora 2048, kv_lora 512, 64 heads, qk_nope 192 / qk_rope 64 / v_head 256, 256 routed experts, top-8, 1 shared expert, `n_group=1`, `topk_group=1`, `route_scale=2.5`, index_topk 2048 / index_n_heads 32 / index_head_dim 128, rope θ=1e6 (interleaved, no YaRN), eps 1e-5, 78 layers (3 dense-FFN + 75 MoE), all layers run sparse DSA.
- Mesh: 8x4 (tp4), sp_axis=0, tp_axis=1.

---

## Data-format / precision map (device vs GPU)

The vLLM golden trace was produced by **`GLM-5.1-FP8`** with native fp8 W8A8 (DeepGEMM) on H200. The
saved tensors are bf16 but encode an **fp8-computed trajectory**. Our device has **no fp8 matmul**.

| Component | Device | GPU (vLLM) | Match? |
|---|---|---|---|
| q_a / q_b / kv_a / o_proj | bf8 (bf16 optional) | fp8 e4m3 DeepGEMM (act per-tok-group-128 × wt 128×128-block, fp32 accum) | precision gap (fp8 vs bf8) |
| routed experts | **bf4** | fp8 e4m3 | precision gap (bf4 vs fp8) |
| shared expert | bf8 | fp8/bf16 | ~match |
| MLA sparse-SDPA (prefill) | **bf16** | **bf16** (flashmla_sparse; KV upconverted) | **MATCH** |
| KV cache | bf8 | fp8_ds_mla → bf16 | precision gap |
| gate matmul / logits | bf16 matmul, fp32 sigmoid | bf16 matmul, fp32 sigmoid | **MATCH** |
| `e_score_correction_bias` | ckpt fp32 → stored **bf16, mean-centered** | fp32 | see gate section |
| indexer | bf16 (logits fp8×fp8→fp32 on GPU) | fp8 scoring | device tracks vLLM 0.98 |
| W_UK / W_UV, norms, embed, lm_head | bf16 / fp32 | bf16 / fp32 | **MATCH** |

Key consequences:
- Our **sparse-SDPA op already matches vLLM** (both bf16). fp8 lives only in the weight GEMMs + indexer scoring.
- **fp8 e4m3 block-scale GEMM ≡ bf16** (measured: CPU-HF teacher-forced vs fp8-GEMM trace and vs bf16 trace agree within ±0.0003/layer). So the GPU's "fp8 trajectory" is effectively a bf16 trajectory; **fp8-sim gains ≈ 0**.
- Device is **bf4-experts + bf8-MLA vs bf16** — the real (small) gap; and, being coarse, it tracks the fp8 trace *slightly better* than a fine bf16 ref does.

---

## References used to validate

| Reference | What it is | Relationship |
|---|---|---|
| **vLLM fp8 trace** | production lineage the device targets (absorbed-MQA + sparse MLA, DeepGEMM fp8) | primary target; non-deterministic run-to-run |
| **HF `glm_moe_dsa` trace** | reference GPU impl (eager, dense non-absorbed MLA); fp8 and bf16 variants; deterministic | independent GPU peer |
| **cpu_v32** (`cpu_deepseek_v32` MLACPU + `TorchMoe`) | GPU-free CPU ref, absorbed MLA → matches device/vLLM | device-validation ref (PCC 1.0 vs `glm_5_1` on L0 dense / L3 MoE) |
| **glm_cpu_hf_ref** | per-layer teacher-forced HF `GlmMoeDsaDecoderLayer` | validates the HF trace (CPU-HF vs GPU-HF: L0 0.9996, L3 0.9932) |
| **injection** | teacher-force GPU routing / topk / sparse-SDPA into the device | isolates routing vs compute |

---

## Results

### Op-level — all faithful
- **Dense L0-2** (attn + DSA + dense FFN, no routing): decoder PCC **0.9995-0.9997**.
- **sparse-SDPA**: input q device↔GPU `0.9995`; output `0.9997 (L0) → 0.9951 (deep)` (grows with depth; op is bf16 both sides — driven by q/kv fp8-vs-bf8 projection drift, not the op).
- **Indexer top-2048 selected keys**: device↔**vLLM** overlap **0.98** (logits PCC `0.9995`); device↔HF `0.73` — but **vLLM↔HF is also 0.73**, so **HF is the outlier**, device faithfully reproduces vLLM. Indexer precision inert (HF bf16-indexer == HF fp8-indexer for the overlap).
- **Gate**: logits device↔GPU **0.99997 @L3**; the selection algorithm (`top8(fp32 sigmoid(logits) + fp32 bias)`) reproduces the GPU's own `expert_ids` at **0.998-1.0** from the GPU's logits → gate compute is exact.
- **MoE FFN reference** (`glm_moe_reference` = `MoEGate` + `TorchExpert`) vs HF `GlmMoeDsaMoE`: **PCC 1.0** (mathematically identical; GLM router == DeepSeek noaux_tc, expert == silu-gated FFN, just packed).

### Block-level (fused `TtPrefillBlock`)
| | dense (L0) | moe (L3) |
|---|---|---|
| real weights (from ttnn cache) | **0.99999996** | **0.99999964** |
| random weights | 0.9997 | 0.9946 |
| vs HF `GlmMoeDsaModel`, real | 0.99996 (full block incl. DSA attn) | — |
| vs HF `GlmMoeDsaModel`, random | 0.9749 (indexer top-2048 knife-edge on untrained scores) | — |

### Chained per-layer (single-shot 5k) — the "0.88 metric"
- device↔vLLM chained decoder PCC: `~1.0` shallow → **trough 0.887 @ L62** → recover `0.960 @ L77`.
- Routing overlap device↔GPU: `0.85 @ L3 → ~0.52 deep → 0.74 @ L77` (mirrors the hidden-PCC curve).
- Peer comparison (all trough @ L62): **device↔vLLM 0.887 ≥ HF↔vLLM 0.883 > device↔HF 0.871** — the device is a peer in the cluster, not an outlier.

### Chunked prefill (5k chunks over 55k)
- L1 `0.9997`, L10 `0.9962`, L78 min `0.8888` (clean `code_debug` 55k trace).
- **chunk-10 anomaly:** the old trace's last chunk collapsed to ~0.84 at deep layers. Ruled out: not the trace (chunk10 L0=0.9997 proves alignment), not cache-fill boundary (bit-identical at different cache sizes), not "last chunk cursed" (chunk is bit-identical whether last or middle). Cause = **intrinsic content** (that chunk had the most repetitive tokens: 736 unique/5120) + depth, tightening the DSA/gate knife-edge. The clean `code_debug` trace removed the collapse (0.84→0.90).

---

## Root cause of the 0.88 chained trough: near-degenerate noaux_tc gate

- **It's MoE expert SELECTION, not compute.** `GLM_INJECT_ROUTING` (force the GPU's expert ids+weights, keep our hidden + our bf4 experts) → **0.97 at every layer**. So attention/experts/hidden are fine once routed correctly.
- **The gate is near-degenerate by design.** noaux_tc (aux-loss-free load balancing) adds a per-expert bias to `sigmoid(logit)` for *selection*; the bias is **anti-correlated with affinity (-0.968 @L3)**, which *flattens* the selection field (~2.6× compression). Result: the 8th-vs-9th expert score gap is `~1.6e-4` (0.05% of range) at L3, up to `5e-3` deep — a knife-edge.
- **Tiny hidden drift → big routing flips.** The device hidden differs from the GPU's by irreducible fp8-vs-bf8/bf4 rounding (hidden PCC `0.80 @L62` after RMSNorm exposes it). The knife-edge amplifies this into ~15% boundary-expert flips @L3, compounding to ~50% @L62.
- **gate_input is the worst-diverging tensor** (`0.80@L62`) — lower than the residual decoder_output (`0.88@L62`), because post-attn RMSNorm strips the common-mode magnitude and exposes the true agreement the gate sees. Chained decoder-PCC *understates* what the gate sees.

### The 0.88 is intrinsic (device-independent proofs)
- **GPU vs GPU (same build, two runs):** routing overlap `0.75-0.95`, chained ceiling `~0.95` — the GPU doesn't reproduce its own routing (fp8 non-determinism, gate-amplified).
- **HF vs vLLM (two GPU impls):** chained trough **0.883 @ L62** — identical to device's 0.887.
- **bf16 CPU-HF chained → 0.47-0.50** (no device at all) — *worse* than the device; and same-code CPU-HF-vs-GPU chained → 0.50. So chained-PCC is intrinsically chaotic for this model.

---

## Levers tried — what moved, what didn't

| Lever | Effect on chained trough (vs vLLM) |
|---|---|
| **inject GPU routing** | 0.886 → **0.971** (the entire lever) |
| inject GPU sparse-SDPA output | +0.056 |
| inject routing + topk (both) | 0.969 (≈ routing alone; topk adds nothing) |
| inject GPU indexer topk **alone** | inert (<0.001) |
| bf16 q/kv projections (HiFi2) | +0.001 (bf16 accum wastes it) |
| bf16 q/kv + HiFi4 + fp32 accum | +0.008 → 0.8899 |
| master-bf16 MLA + HiFi4 | 0.891 (best) |
| KV cache bf16 | inert (overshoots vLLM's fp8 KV) |
| fp32 gate matmul | inert deep |
| `GLM_GATE_BIAS_FP32` | **BROKEN — do not use** (native fp32 bias tile mishandled by `moe_grouped_topk` → routes to degenerate {0..7} → crashes chained 0.886 → 0.525) |
| gate bias **mean-center + bf16** (default) | **CORRECT / required** (0.974→0.998 vs fp32; raw bf16 bias without mean-sub is catastrophic 0.23-0.83) |
| expert fp8-sim (e4m3 block-scale) | inert (bf4 ≡ bf8 ≡ fp8 ≡ bf16 for experts) |
| overall fp8-sim | **dead end** (fp8-GEMM ≡ bf16) |
| bf8 routed experts | **blocked** (L1 overflow +4%) |

Precision ladder (device vs vLLM, fp8 source): bf8 `0.8822` → bf16 q/kv HiFi2 `0.8834` → bf16+HiFi4+fp32acc `0.8899` → master-bf16-MLA+HiFi4 `0.891` → +kvpe-bf16 `0.8908` (inert). **Ceiling ~0.89** with bf4 experts.

---

## Gate-bias precision note (surfaced again by the block test)

The device stores `e_score_correction_bias` as **bf16** (even in DEVICE_FP32, which only makes the
*logits* fp32). GLM's real bias is a **large DC offset with a tiny inter-expert spread** (e.g. layer 3
≈ 34.5 ± 0.14; other layers ≈ 10.16 ± 0.04), and that spread *decides* the top-8. bf16's ULP near
those magnitudes (~0.06-0.25) is coarser than the spread, so:
- The **mean-centering fix** (subtract the DC offset before bf16 storage) is essential — it recovers
  the spread (routing 0.974→0.998 vs fp32). Without it, routing is 0.23-0.83 (catastrophic).
- Residual bias-precision loss after mean-centering is tiny (~0.002 deep); the deep divergence is the
  hidden drift, not the bias.
- For a *reference* comparison you must bf16-match the bias too (else the fp32 ref routes differently:
  fp32-vs-bf16 top-8 overlap was 0.38 for the raw bias). This is why the block MoE reference quantizes
  the bias to bf16.

Possible improvement (follow-up): keep the bias fp32 into `moe_grouped_topk` in DEVICE_FP32 mode
(currently the op mishandles a native fp32 bias tile). Small win (~0.002 deep), but tidies the path.

---

## Known issues / follow-ups
1. **bf8 routed experts — L1 overflow** (+63 KB / 4% over 1.5 MB in `UnifiedRoutedExpertFfn`); needs a CB/subblock program-config tweak. Until then experts are bf4 (ceiling 0.891 vs vLLM).
2. **Device gate bf16 bias** in DEVICE_FP32 (mean-centering makes it correct; a true-fp32-bias path in `moe_grouped_topk` is a tidy-up, small).
3. **Chained-PCC as a CI gate is inappropriate** for GLM — it's below 0.95 for every faithful impl incl. reference GPUs. Certify by next-token / teacher-forced / component / block PCC.

---

## How the references are wired (plumbing, for reproduction)
- **Injection knobs:** `GLM_INJECT_ROUTING`, `GLM_INJECT_TOPK`, `GLM_INJECT_SPARSE_SDPA` (teacher-force the GPU's routing / indexer keys / attention output).
- **Precision knobs:** `GLM_MLA_QKV_BF16`, `GLM_MLA_WUV_BF16`, `GLM_MLA_FP32_ACC` (HiFi4+fp32acc), `GLM_KVPE_BF16`, `GLM_GATE_FP32_MATMUL`. (`GLM_GATE_BIAS_FP32` exists but is broken — leave off.)
- **Dumps:** `GLM_DUMP_GATE_LOGITS`, `GLM_DUMP_GATE_INPUT`, `GLM_DUMP_SDPA`, `GLM_DUMP_TOPK_INDICES`, `GLM_DUMP_DEVICE_SNAPSHOTS` (deterministic device forward → regrade offline vs any trace).
- **Weight source:** `GLM51_REPO=/mnt/models/zai-org/GLM-5.1` (bf16 master) or `.../GLM-5.1-FP8` (fp8, dequant-handled). The fp8 loader dequantizes to bf16 = the GPU's exact fp8 weight values in 16 bits.
- **Block test / CPU ref:** device loads real weights from the ttnn tensorbin cache; the CPU reference (`reference/glm_5_1`) composes `SparseMLAReference` (DSA-MLA) + per-expert noaux_tc MoE, verified == HF `GlmMoeDsaMoE` at PCC 1.0.

___
*** All PCCs are per-layer output ***

#### PCC between GPU traces.
<img width="1435" height="719" alt="Image" src="https://github.com/user-attachments/assets/2e73c4c5-20da-4c33-ae8f-ae59b7e6e54b" />

#### Injecting indexer-topk + expert routing from GPU trace to our model run
<img width="1443" height="666" alt="Image" src="https://github.com/user-attachments/assets/8a42bb4f-ce11-413c-a5dc-d01840f82e4d" />

#### Prefill transformer chunked run, 11th chunk bad
<img width="1437" height="819" alt="Image" src="https://github.com/user-attachments/assets/1356595b-71a8-432b-a2bf-00fb8ca5e7a7" />

#### PCC between our run ttnn vs VLLM reference vs HF reference (HF reference simulates sparse attn with dense)
<img width="1444" height="658" alt="Image" src="https://github.com/user-attachments/assets/821c5e00-38e3-45ed-96d5-d1f08c1fb3a9" />

<hr>

CPU fallback of whole MLA - no difference (0.002 pcc bump, meaningless)

**Comment by nmilicevicTT** [comment]
Expert weights bf8 vs bf4. Using bf8 lifts pcc meaningfully. This is for GLM5.2 but the pcc trend looks the same so it applies. Teacher-forced means we take gpu outputs and use them as inputs to our layers. 
<img width="1444" height="536" alt="Image" src="https://github.com/user-attachments/assets/1ea597ca-11e2-4b84-a75c-44953f708712" />

**Comment by mbezuljTT** [comment]
Please double check with decode if they are using bfp4 for routed expert weights.

**Comment by ipotkonjak-tt** [comment]
Decode is using bfp4 weights for GLM 5.2
